tomcat9/src/test/java/com/linecorp/armeria/server/tomcat/TomcatServiceAccessLogTest.java

```diff
@@ -0,0 +1,114 @@
+ */
+
+package com.linecorp.armeria.server.tomcat;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.CharArrayWriter;
+import java.io.File;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.catalina.startup.Tomcat;
+import org.apache.catalina.valves.AbstractAccessLogValve;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+class TomcatServiceAccessLogTest {
+    private static Tomcat tomcat;
+
+    private static final FakeAccessLogValve accessLogValve = FakeAccessLogValve.create();
+
+    public static final class FakeAccessLogValve extends AbstractAccessLogValve {
+        static FakeAccessLogValve create() {
+            final FakeAccessLogValve fakeAccessLogValve = new FakeAccessLogValve();
+            fakeAccessLogValve.setPattern("%D");
+            return fakeAccessLogValve;
+        }
+
+        private List<String> logs = Collections.synchronizedList(new ArrayList<>());
+
+        private FakeAccessLogValve() {
+        }
+
+        @Override
+        protected void log(CharArrayWriter message) {
+            logs.add(message.toString());
+        }
+
+        String popLog() {
+            return logs.remove(0);
+        }
+
+        void reset() {
+            logs.clear();
+        }
+    }
+
+    @RegisterExtension
+    static final ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            // Prepare Tomcat instances.
+            tomcat = new Tomcat();
+            tomcat.setPort(0);
+            tomcat.setBaseDir("build" + File.separatorChar +
+                              "tomcat-" + TomcatServiceAccessLogTest.class.getSimpleName());
+            TomcatUtil.engine(tomcat.getService(), "bar");
+            tomcat.getEngine().getPipeline().addValve(accessLogValve);
+
+            // Start the Tomcats.
+            tomcat.start();
+            // Bind them to the Server.
+            sb.serviceUnder("/no-webapp/", TomcatService.of(tomcat));
+        }
+    };
+
+    @AfterAll
+    static void destroyTomcat() throws Exception {
+        if (tomcat != null) {
+            tomcat.stop();
+            tomcat.destroy();
+        }
+    }
+
+    @BeforeEach
+    void beforeEach() {
+        accessLogValve.reset();
+    }
+
+    @Test
+    void haveCorrectProcessingTime() throws Exception {
+        try (CloseableHttpClient hc = HttpClients.createMinimal()) {
```

Comment:
nit: Why not use Armeria's `WebClient`?

Reply:
ah due to copy from existing UnmanagedTomcatServiceTest.
Update to armeria one^^